physical/raft/raft_autopilot.go

```diff
@@ -0,0 +1,718 @@
+	ret := make(map[raft.ServerID]*autopilot.ServerStats)
+
+	d.l.RLock()
+	followerStates := d.followerStates
+	d.l.RUnlock()
+
+	followerStates.l.RLock()
+	defer followerStates.l.RUnlock()
+
+	now := time.Now()
+	for id, followerState := range followerStates.followers {
+		ret[raft.ServerID(id)] = &autopilot.ServerStats{
+			LastContact: now.Sub(followerState.LastHeartbeat),
+			LastTerm:    followerState.LastTerm,
+			LastIndex:   followerState.AppliedIndex,
+		}
+	}
+
+	leaderState, _ := d.fsm.LatestState()
+	ret[raft.ServerID(d.localID)] = &autopilot.ServerStats{
+		LastTerm:  leaderState.Term,
+		LastIndex: leaderState.Index,
+	}
+
+	return ret
+}
+
+// KnownServers is called by the autopilot library to know the status of each
+// node in the raft cluster. If the application thinks that certain nodes left,
+// it is here that we let the autopilot library know of the same.
+func (d *Delegate) KnownServers() map[raft.ServerID]*autopilot.Server {
+	d.l.RLock()
+	future := d.raft.GetConfiguration()
+	if err := future.Error(); err != nil {
+		d.logger.Error("failed to get raft configuration when computing known servers", "error", err)
+		d.l.RUnlock()
+		return nil
+	}
+
+	servers := future.Configuration().Servers
+	serverIDs := make([]string, 0, len(servers))
+	for _, server := range servers {
+		serverIDs = append(serverIDs, string(server.ID))
+	}
+	followerStates := d.followerStates
+	d.l.RUnlock()
+
+	followerStates.l.RLock()
+	defer followerStates.l.RUnlock()
+
+	ret := make(map[raft.ServerID]*autopilot.Server)
+	for id, state := range d.followerStates.followers {
+		// If the server is not in raft configuration, even if we received a follower
+		// heartbeat, it shouldn't be a known server for autopilot.
+		if !strutil.StrListContains(serverIDs, id) {
+			continue
+		}
+
+		server := &autopilot.Server{
+			ID:          raft.ServerID(id),
+			Name:        id,
+			RaftVersion: raft.ProtocolVersionMax,
+			Ext:         d.autopilotServerExt(state.DesiredSuffrage),
+		}
+
+		switch state.IsDead {
+		case true:
+			d.logger.Debug("informing autopilot that the node left", "id", id)
+			server.NodeStatus = autopilot.NodeLeft
+		default:
+			server.NodeStatus = autopilot.NodeAlive
+		}
+
+		ret[raft.ServerID(id)] = server
+	}
+
+	// Add the leader
+	ret[raft.ServerID(d.localID)] = &autopilot.Server{
+		ID:          raft.ServerID(d.localID),
+		Name:        d.localID,
+		RaftVersion: raft.ProtocolVersionMax,
+		NodeStatus:  autopilot.NodeAlive,
+		Ext:         d.autopilotServerExt("voter"),
+	}
+
+	return ret
+}
+
+// RemoveFailedServer is called by the autopilot library when it desires a node
+// to be removed from the raft configuration. This function removes the node
+// from the raft cluster and stops tracking its information in follower states.
+// This function needs to return quickly. Hence removal is performed in a
+// goroutine.
+func (d *Delegate) RemoveFailedServer(server *autopilot.Server) {
+	go func() {
```

Comment:
The possibility of accumulating goroutines trying to do this when there's a failure makes me a little uneasy.  I note on the Consul side they're just sending a gossip message in the equivalent method.  I wonder if we should do something more, like have a map keyed by nodeid to keep track of ongoing removals?

Reply:
Something like this? https://github.com/hashicorp/vault/pull/11019
The problem I see there is I am not sure how to clean the entries in the map.

Reply:
Yeah, that's the sort of thing I had in mind.  I don't think a sync.Map is adequate to manage concurrency here - I mean in practice it'll probably be, because we don't expect RemoveFailedServer to be invoked twice in very quick succession, but even so I'd rather we not commit inappropriate patterns that others might copy.  Using sync.Map will prevent the race detector from complaining, but suppose that we had two more or less simultaneous calls.  Each would call Load, see that there's no entry for their serverID, and proceed in parallel instead of one returning.  This would be a case where you'd need a mutex to get sensible behaviour.

As to cleaning map entries, I would just remove the map entry whenever the goroutine returns, e.g. in a defer.  Though you'll need to grab the lock for that removal too.

Please note that this is a pretty low-priority concern of mine.  I think it makes sense to address it, but I'd be totally fine with you ignoring it for now in favour of testing or whatever.  I don't expect this'll be an issue except possibly in extreme scenarios.

Reply:
Resolved in https://github.com/hashicorp/vault/pull/11019